src/lexer.rs

```diff
@@ -0,0 +1,319 @@
+//! Lexer for semver ranges.
+//!
+//! Breaks a string of input into an iterator of tokens that can be used with a parser.
+//!
+//! This should be used with the [`parser`] module.
+//!
+//! [`parser`]: ../parser/index.html
+//!
+//! # Examples
+//!
+//! Example without errors:
+//!
+//! ```rust
+//! use semver_parser::lexer::{Lexer, Token};
+//!
+//! let mut l = Lexer::new("foo 123 *");
+//!
+//! assert_eq!(Some(Ok(Token::AlphaNumeric("foo"))), l.next());
+//! assert_eq!(Some(Ok(Token::Whitespace(3, 4))), l.next());
+//! assert_eq!(Some(Ok(Token::Numeric(123))), l.next());
+//! assert_eq!(Some(Ok(Token::Whitespace(7, 8))), l.next());
+//! assert_eq!(Some(Ok(Token::Star)), l.next());
+//! assert_eq!(None, l.next());
+//! ```
+//!
+//! Example with error:
+//!
+//! ```rust
+//! use semver_parser::lexer::{Lexer, Token, Error};
+//!
+//! let mut l = Lexer::new("foo / *");
+//!
+//! assert_eq!(Some(Ok(Token::AlphaNumeric("foo"))), l.next());
+//! assert_eq!(Some(Ok(Token::Whitespace(3, 4))), l.next());
+//! assert_eq!(Some(Err(Error::UnexpectedChar('/'))), l.next());
+//! ```
+
+use std::str;
+use self::Token::*;
+use self::Error::*;
+
+macro_rules! scan_while {
+    ($slf:expr, $start:expr, $first:pat $(| $rest:pat)*) => {{
+        let mut __end = $start;
+
+        loop {
+            if let Some((idx, c)) = $slf.one() {
+                __end = idx;
+
+                match c {
+                    $first $(| $rest)* => $slf.step(),
+                    _ => break,
+                }
+
+                continue;
+            } else {
+                __end = $slf.input.len();
+            }
+
+            break;
+        }
+
+        __end
+    }}
+}
+
+/// Semver tokens.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum Token<'input> {
+    /// `=`
+    Eq,
+    /// `>`
+    Gt,
+    /// `<`
+    Lt,
+    /// `<=`
+    LtEq,
+    /// `>=`
+    GtEq,
+    /// '^`
+    Caret,
+    /// '~`
+    Tilde,
+    /// '*`
+    Star,
+    /// `.`
+    Dot,
+    /// `,`
+    Comma,
+    /// `-`
+    Hyphen,
+    /// `+`
+    Plus,
+    /// '||'
+    Or,
+    /// any number of whitespace (`\t\r\n `) and its span.
+    Whitespace(usize, usize),
+    /// Numeric component, like `0` or `42`.
+    Numeric(u64),
+    /// Alphanumeric component, like `alpha1` or `79deadbe`.
+    AlphaNumeric(&'input str),
+}
+
+impl<'input> Token<'input> {
+    /// Check if the current token is a whitespace token.
+    pub fn is_whitespace(&self) -> bool {
+        match *self {
+            Whitespace(_, _) => true,
```

Comment:
this isn't really an issue but I thought I'd mention it; you can also do

    Whitespace(..) => true,

here, which is *ever so slightly* more clear, personally.

i don't think this change needs to be made but if you agree and happen to do it, no big deal :)

Reply:
Fixed.